text/0064-refer-to-types-in-empty-namespace.md

```diff
@@ -0,0 +1,159 @@
+namespace.
+The implicit-fallback change deals with referencing a typename in the empty
+namespace where there is no collision with a typename in the active namespace.
+The explicit-`::` change deals with referencing a typename in the empty
+namespace where there _is_ a collision with a typename in the active namespace.
+
+This issue has been encountered "in the wild" by multiple people, according to
+a [comment on cedar#579](https://github.com/cedar-policy/cedar/issues/579#issuecomment-2077482637).
+
+## Detailed design
+
+### Change 1: Fully qualified syntax for referring to typenames in the empty namespace
+
+The syntax `::id`, where `id` is an unqualified typename, now explicitly refers
+to the `id` in the empty namespace.
+
+This applies:
+* regardless of if there is an `id` defined in the active namespace or not
+* regardless of if there is an `id` defined in the empty namespace or not (if not, this is an error)
+* regardless of whether `id` is the name of a common type or an entity type, in
+either the empty and/or active namespace
+* in both the human (RFC 24) and JSON schema syntaxes
+
+This does not apply:
+* to primitive and extension types. From the user's perspective, those live in
+the `__cedar` namespace and not in the empty namespace. `::String` refers to an
+entity or common type `String` defined in the empty namespace, while
+`__cedar::String` refers to the builtin primitive type `String`.
+* to "names" like `::A::B` -- that syntax remains an error. `A::B` is already an
+unambiguous, fully-qualified name, and will never have any namespace implicitly
+prepended.
+
+### Change 2: Implicitly falling back on the empty namespace
+
+When the schema refers to an unqualified typename (e.g, `id`) which doesn't
+exist in the active namespace, we should fall back to looking for it in the
+empty namespace.
+Only if it exists in neither the active nor empty namespace should this be an
+error.
+
+E.g., in the example at the top of this RFC, consider the type of the `id`
+attribute of `Demo::User`.
+Since there is no `id` type defined in namespace `Demo`, this is an error today.
+With this change, when we see there is no `Demo::id`, we implicitly look for
+`id` in the empty namespace, and resolving `id` to be the common type `::id`
+defined to be `String` in the empty namespace.
+The user could, of course, write `::id` directly, but with this change, they
+do not have to, in the common case where there is no collision (where `id` is
+not actually defined in both the `Demo` and empty namespaces).
+
+This applies to both entity and common types defined in the empty namespace.
+
+This applies to both the human (RFC 24) and JSON schema syntaxes.
+
+This is reminiscent of, but not dependent on, the existing similar behavior for
+primtiive and extension types, where we first check for an entity or common type
+`String` in the active namespace before falling back on resolving to
+`__cedar::String`.
+
+## Drawbacks
+
```

Comment:
Another possible drawback is that syntax `::Foo` is unprecedented for existing languages, as far as I understand it. So customers would not reach for it intuitively.

It also looks a little strange to me to see `owner: ::User`; that extra space is now really important. You might want to specially deal in the parser with the mistake of writing `owner:::User`

Reply:
if `owner:User` and `owner:Foo::User` are valid without a space after the first `:`, then you could argue `owner:::User` should be parsable.  I found that this seems to parse just fine without extra spaces:

```
entity P;
action b;
action"a"in[b]appliesTo{principal:P};
```

I still like `__::User` since `__` is already a special prefix and can't collide with a user defined namespace.  `owner:__::User` is perfectly readable.

Reply:
> `__` is already a special prefix and can't collide with a user defined namespace

The current draft of [RFC 52](https://github.com/cedar-policy/rfcs/pull/52) proposes reserving just `__cedar` and not all namespaces starting with `__`.

Reply:
I misremembered the ident rules and thought `__` wouldn't be valid today and would be safe.  Thanks for the clarification.

```
IDENT ::= ['_''a'-'z''A'-'Z']['_''a'-'z''A'-'Z''0'-'9']* - RESERVED
```

Reply:
In my opinion, the concerns in this thread are valid and sufficient to make us prefer a syntax other than `::User`.  In particular, I agree that `owner:::User` would have to be legal and that it's confusing to mentally parse.

I'll come up with an alternative and revise this RFC.